tls.go

```diff
@@ -80,3 +85,64 @@ func readclientKeyCertificate(certFile, keyFile string) ([]byte, []byte, error)
 
 	return clientCert, clientKey, nil
 }
+
+// setupClientCertificate uses the provided tls.Certificate and caCert bytes to create/modify tls.Config
+func setupClientCertificate(clientCertWithKey tls.Certificate, caChain []*x509.Certificate, tlsConfig *tls.Config) (*tls.Config, error) {
+	if clientCertWithKey.PrivateKey == nil {
+		return nil, fmt.Errorf("supplied certificate does not have a private key")
+	}
+
+	if len(caChain) == 0 {
+		return nil, fmt.Errorf("no ca certificate(s) supplied")
+	}
+
+	if tlsConfig == nil {
+		tlsConfig = &tls.Config{}
+	}
+
+	// If required, instantiate CA certificate pool
+	if tlsConfig.RootCAs == nil {
+
+		caCertPool, err := x509.SystemCertPool()
+		if err != nil {
+			return nil, fmt.Errorf("failed to obtain system CA pool: %s", err)
+		}
+
+		tlsConfig.RootCAs = caCertPool
+	}
+
+	for _, cert := range caChain {
+		tlsConfig.RootCAs.AddCert(cert)
+	}
+
+	// Append client certificate to config
+	tlsConfig.Certificates = append(tlsConfig.Certificates, clientCertWithKey)
+
+	return tlsConfig, nil
+}
+
+// ParseCAChain takes a file of PEM encoded things and returns the CERTIFICATEs in order
+// taken and adapted from crypto/tls
+func ParseCAChain(caCert []byte) ([]*x509.Certificate, error) {
```

Comment:
Does this have to be public? If it's a helper function (I can't see any references / uses here), is it maybe beyond the scope of the package? Does it need tests?

Reply:
it's copy pasted from crypto/tls.